graphics/src/widget/canvas/frame.rs

```diff
@@ -295,46 +297,133 @@ impl Frame {
+                alignment::Vertical::Center => {
+                    text.position.y + total_height / 2.0
+                }
+                alignment::Vertical::Bottom => text.position.y + total_height,
+            }
+        };
+
+        for run in layout.iter() {
+            for glyph in run.glyphs.iter() {
+                let start_x = translation_x + glyph.x + glyph.x_offset;
+                let start_y = translation_y + glyph.y_offset - text.size;
+
+                let offset = Vector::new(start_x, start_y);
+
+                let mut swash_cache = cache.swash.borrow_mut();
+
+                if let Some(commands) =
+                    swash_cache.get_outline_commands(glyph.cache_key)
+                {
+                    let glyph = Path::new(|path| {
+                        use cosmic_text::Command;
+
+                        for command in commands {
+                            match command {
+                                Command::MoveTo(p) => {
+                                    path.move_to(
+                                        Point::new(p.x, -p.y) + offset,
+                                    );
+                                }
+                                Command::LineTo(p) => {
+                                    path.line_to(
+                                        Point::new(p.x, -p.y) + offset,
+                                    );
+                                }
+                                Command::CurveTo(control_a, control_b, to) => {
+                                    path.bezier_curve_to(
+                                        Point::new(control_a.x, -control_a.y)
+                                            + offset,
+                                        Point::new(control_b.x, -control_b.y)
+                                            + offset,
+                                        Point::new(to.x, -to.y) + offset,
+                                    );
+                                }
+                                Command::QuadTo(control, to) => {
+                                    path.quadratic_curve_to(
+                                        Point::new(control.x, -control.y)
+                                            + offset,
+                                        Point::new(to.x, -to.y) + offset,
+                                    );
+                                }
+                                Command::Close => {
+                                    path.close();
+                                }
+                            }
+                        }
+                    });
+
+                    self.fill(&glyph, text.color);
+                } else {
+                    // TODO: Raster image support for `Canvas`
+                    let [r, g, b, a] = text.color.into_rgba8();
+
+                    swash_cache.with_pixels(
+                        glyph.cache_key,
+                        cosmic_text::Color::rgba(r, g, b, a),
```

Comment:
I did some profiling and the bitmap pixels -> lyon tessellation is a huge performance killer. Even on release mode, I'm getting primitive generation times for size 80 text in the `8ms` range. This drops to `~400μs` when removing the emoji.

Maybe we add a disclaimer to the `fill_text` docs until there is better support for rastered images w/ transformations.

Reply:
I ported the new example to `master`, but it's near impossible compare performance differences between the old / new `fill_text` since previously this was GPU rendered text and now it's tessellated by CPU. New changes are exposed in `primitive generation` numbers, but previously the impact would be hidden in `render`. Both maintain 60fps in release mode.

Obviously the canvas will most likely sit behind a `canvas::Cache` and in most workflows, the text is fairly static outside applying transformations.

Reply:
Yes, as I mentioned it's pretty inefficient!

The bitmap support is just a nice to have. For vectorial text, you'll want to deal with fonts that have outline information, so I don't think it's a big deal.

That said, we can probably skip the generic tessellation algorithm and hardcode the triangles ourselves.

Reply:
Also, important to note that once we remove `glyph-brush`, we will be able to compose primitives much more efficiently. We may be able to use the GPU glyph atlas approach for text that isn't transformed for better readability and performance.

Lots of doors will open once we get that bottleneck out of the way.

Reply:
Makes sense!

> That said, we can probably skip the generic tessellation algorithm and hardcode the triangles ourselves.

That would be nice. Even though tessellating rectangles should be very cheap, I'd imagine there's still some overhead with `lyon` that we can get around building the mesh ourselves.

Reply:
I believe it's as cheap as any other path, since we are using the generic path algorithm and not `fill_rectangle` (which only works with axis-aligned rectangles).